osu.Game/Rulesets/Difficulty/PerformanceCalculator.cs

```diff
@@ -21,14 +21,14 @@ public abstract class PerformanceCalculator
 
         protected double TimeRate { get; private set; } = 1;
 
-        protected PerformanceCalculator(Ruleset ruleset, WorkingBeatmap beatmap, ScoreInfo score)
+        protected PerformanceCalculator(Ruleset ruleset, WorkingBeatmap beatmap, ScoreInfo score, DifficultyAttributes attributes = null)
```

Comment:
Ideally instead of having to jam the `DifficultyAttributes` through here, `PerformanceCalculator` should interface with `BeatmapDifficultyManager` directly, therefore directly using its cache. That direction might therefore lead to it becoming a `Component` or otherwise a part of the drawable hierarchy to leverage DI.

I'm not 100% on this one - @smoogipoo, how do you see this going forward?

Reply:
I don't like that path, because it kinda removes the "headless" capability of this which is used in server-side components. Don't want to instantiate a whole `GameHost` to use this.

Reply:
Ah, right, forgot about that... Maybe some sort of DI-able wrapper over `PerfomanceCalculator`, then? Not sure, but I really dislike the current form of having to have the user of `PerformanceCalculator` also require using `BeatmapDifficultyManager` and manually passing the results of one to the other to get the cache gains.

Reply:
not sure I follow the reasoning here. wouldn't that just add another layer of indirection with very little gain ? (chances are, it probably will only be used here)

Reply:
For one, if the direction is to use difficulty attributes for pp calculation then `WorkingBeatmap` should be replaced entirely with non-nullable `DifficultyAttributes`, which I'd be more fine with than this current version. I'm not sure whether having a performance calculator necessitate always instantiating a difficulty attribute calculator to be able to *do anything* with it is good design but I suppose it's no worse than what we have currently.